File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        opening = []
        for cur in s:
            if cur == "(" or cur == "{" or cur == "[":
                opening.append(cur)
            elif len(opening) > 0: 
                if (cur == ")" and opening[-1] == "(") or (cur == "]" and opening[-1] == "[") or (cur == "}" and opening[-1] == "{"):
                    opening.pop()
                else:
                    return False    
            else:
                return False
        
        if len(opening) > 0:
            return False
        return True

File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        LinkedList<Character> stack = new LinkedList<>();
        for (int i = 0; i < s.length(); i++) {
           char cur = s.charAt(i);
            if (cur == '(' || cur == '[' || cur == '{'){
                stack.push(cur);
            } else {
                if (stack.isEmpty()) {
                    return false;
                } else if (cur == ')' && stack.pop() != '(') {
                    return false;
                } else if (cur == ']' && stack.pop() != '[') {
                    return false;
                } else if (cur == '}' && stack.pop() != '{') {
                    return false;
                }
            }
        }
        if (stack.isEmpty()) {
            return true;
        }
        return false;
    }
}